encoders/firebase-decoders-json/src/json/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContext.java

```diff
@@ -67,6 +80,8 @@
       return decodePrimitive(classToken);
     } else if (isSingleValue(classToken)) {
       return decodeSingleValue(classToken);
+    } else if (isPreDefinedObject(classToken)) {
+      return decodePreDefinedObject(classToken);
```

Comment:
Looks like these 2 methods do more than one thing, consider splitting their functionality in to map and collection specific stuff, context: https://en.wikipedia.org/wiki/Single-responsibility_principle